test/Microsoft.TestPlatform.AcceptanceTests/Extension/RunnerCompatibilityDataSource.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Reflection;
+
+namespace Microsoft.TestPlatform.AcceptanceTests;
+
+/// <summary>
+/// A data source that provides every version of runner.
+/// 
+/// When that adds up to no configuration exception is thrown.
+/// </summary>
+public class RunnerCompatibilityDataSource : TestDataSource<RunnerInfo>
+{
+    private readonly CompatibilityRowsBuilder _builder;
+
+    public RunnerCompatibilityDataSource(
+        string runnerFrameworks = AcceptanceTestBase.DEFAULT_HOST_NETFX_AND_NET,
+        string runnerVersions = AcceptanceTestBase.LATEST_TO_LEGACY,
+        string hostFrameworks = AcceptanceTestBase.DEFAULT_HOST_NETFX_AND_NET)
+    {
+        // TODO: We actually don't generate values to use different translation layers, because we don't have a good way to do
+        // that right now. Translation layer is loaded directly into the acceptance test, and so we don't have easy way to substitute it.
+
+        _builder = new CompatibilityRowsBuilder(
+            runnerFrameworks,
+            runnerVersions,
+            hostFrameworks,
+            // host versions
+            AcceptanceTestBase.LATEST,
+            // adapter versions
+            AcceptanceTestBase.LATESTSTABLE,
+            // adapters
+            AcceptanceTestBase.MSTEST);
+
+        // Do not generate the data rows here, properties (e.g. DebugVSTestConsole) are not populated until after constructor is done.
+    }
+
+    public bool DebugVSTestConsole { get; set; }
```

Comment:
Not sure if this will bring much but I see this set of properties appears in multiple data sources, is it worth having an intermediate type? If we don't foresee much change going forward I don't mind the "copy/paste" state.

Reply:
I was thinking about inheritance and not composition here.

```csharp
public abstract class TestDataSourceBase : TestDataSource<RunnerInfo>
{
	public bool DebugVSTestConsole { get; set; }
    public bool DebugTestHost { get; set; }
    public bool DebugDataCollector { get; set; }
    public bool NoDefaultBreakpoints { get; set; } = true;
}

public class MSTestCompatibilityDataSource : TestDataSourceBase
{
...
}
```

But you are right that we would need to do something similar for attribute so that's not ideal.

For the composition, the only thing that actually makes sense would be to reuse the `DebugInfo` type directly instead of the 4 properties on the various `TestDataSource` or `Attribute` types.